opentreemap/treemap/udf.py

```diff
@@ -1094,43 +1106,270 @@ class UDFModel(UserTrackable, models.Model):
+                self.instance.dirty_collection_udfs.add(key)
+                del self.collection_fields[key]
+            else:
+                super(UDFModel.UdfsProxy, self).__delitem__(key)
+
+                hstore_attr = getattr(self.instance, self._field_name)
+                if isinstance(hstore_attr, dict):
+                    del hstore_attr[key]
+
+        # The automated tests require filling in all the collection methods
+
+        def get(self, key, default):
+            try:
+                return self[key]
+            except KeyError:
+                return default
+
+        def has_key(self, key):
+            try:
+                self[key]
+                return True
+            except KeyError:
+                return False
+
+        def iteritems(self):
+            for udfd in self.instance.get_user_defined_fields():
+                v = None
+                try:
+                    v = self.__getitem__(udfd.name)
+                    if v is not None:
+                        yield udfd.name, v
+                except KeyError:
+                    pass
+
+        def iterkeys(self):
+            for k, v in self.iteritems():
+                yield k
+
+        def itervalues(self):
+            for k, v in self.iteritems():
+                yield v
+
+        def items(self):
+            return [i for i in self.iteritems()]
+
+        def keys(self):
+            return [k for k in self.iterkeys()]
+
+        def values(self):
+            return [v for v in self.itervalues()]
+
     def __init__(self, *args, **kwargs):
+        # Model doesn't know anything about `'udfs'`,
+        # so take it out before `super`.
+        udfs_kwarg = kwargs.pop('udfs', {})
+
         super(UDFModel, self).__init__(*args, **kwargs)
-        # Collection UDF audits are handled by the UDFCollectionValue class
+
+        self._setup_udf_model_type()
+
+        hstore_udfs_kwarg = copy.deepcopy(getattr(self, 'hstore_udfs', {}))
+        self._setup_udfs(udfs_kwarg, hstore_udfs_kwarg)
+
+        # Collection UDF audits are handled by the
+        # UserDefinedCollectionValue class
         self._collection_field_names = {udfd.canonical_name
                                         for udfd in self.collection_udfs}
         # This is the whole reason for keeping _do_not_track
         # in addition to the do_not_track class property.
         self._do_not_track |= self.do_not_track | self._collection_field_names
-        self.populate_previous_state()
 
-        self.dirty_collection_udfs = {}
+        # It is the leaf class' responsibility to call
+        # `self.populate_previous_state()` after initialization is
+        # otherwise complete.
+
+        self.dirty_collection_udfs = set()
+
+    def _setup_udf_model_type(self):
+        '''
+        Model type used in searching for UserDefinedFieldDefinition objects
+        related to self.
+
+        Cache it in self.__dict__ for use by __getattr__, to avoid
```

Comment:
The usage moved to `_get_user_defined_fields_from_dict`, called from `self.get_user_defined_fields`, but I forgot to update the comment. Will do so asap.